pkg/vacuum/vacuum.go

```diff
@@ -275,25 +305,72 @@ func (e *Engine) Run(ctx context.Context) {
 		}
 		numberVacuumConnections.Set(0)
 	}()
-	// we limit ourselves to batches of 1000 chunks
-	// since we already have the advisory lock, continue to vacuum batches as needed until none left
-	for {
-		chunks, err := e.listChunks(ctx, con)
+
+	// okay, we have the advisory lock, let's do some work...
+
+	chunksToFreeze := workload{
+		name:         "compressed chunks to freeze",
+		query:        sqlListChunksToFreeze,
+		scaleWorkers: true,
+		finished:     false,
+		failed:       false,
+	}
+	chunksMissingStats := workload{
+		name:         "compressed chunks missing stats",
+		query:        sqlListChunksMissingStats,
+		scaleWorkers: false,
+		finished:     false,
+		failed:       false,
+	}
+
+	for i := 0; true; i++ {
+		// continue looping until both workloads either finished or failed
+		if (chunksToFreeze.finished || chunksToFreeze.failed) && (chunksMissingStats.finished || chunksMissingStats.failed) {
+			break
+		}
+
+		// every 3rd iteration is chunksMissingStats
+		w := &chunksToFreeze
+		if i%3 == 0 {
+			w = &chunksMissingStats
+		}
+
+		// if this workload already finished, don't continue working it, work the other
+		if w.finished {
+			continue
```

Comment:
doesn't this mean chunksMissingStats can indefinitely starve out chunksToFreeze after chunksToFreeze.finish becomes true?

Reply:
yes. dangit.